frontend/src/frontend_benchmark/source_generator.py

```diff
@@ -0,0 +1,185 @@
+        string = self.callgraph.format_function(function_name)
+        with open(path, 'a') as f:
+            f.write(string + '\n')
+
+    def write_makefile(self, function_files: Collection[str]) -> None:
+        makefile_path = self.append_path_to_output_dir('Makefile')
+        with open(makefile_path, 'w') as f:
+            dependencies = self.get_object_files_to_c_files_mapping(function_files)
+            obj_files = ' '.join(dependencies.keys())
+            string = f'{self.benchmark_name}: {obj_files}\n'
+            string = f'{string}\tgcc -o {self.benchmark_name} {obj_files} -O0\n\n'
+            for obj_file, c_file in dependencies.items():
+                string += f'{obj_file}: {c_file}\n'
+                string += f'\tgcc -c -o {obj_file} {c_file} -O0\n\n'
+            string += f'clean:\n\trm *.o {self.benchmark_name}\n'
+            f.write(string)
+
+    def get_object_files_to_c_files_mapping(self, c_files: Collection[str]) -> Dict[str, str]:
+        result = {}
+        for func_file in c_files:
+            obj_file = re.sub(r'\.c', r'.o', func_file)
+            result[obj_file] = func_file
+        result['main.o'] = 'main.c'
+        return result
+
+
+class FileFunctionMapper:
+    def __init__(self, callgraph):
+        self.function_files = set()
+        self.callgraph = callgraph
+
+    def group_functions(self, splits: int = None):
+        result = self._group_functions_without_splits()
+        if splits:
+            result = self._group_functions_with_splits(splits, result)
+        return result
+
+    def _group_functions_without_splits(self) -> Dict[str, Set[int]]:
+        from collections import deque, defaultdict
+        visited_functions = set()
+        function_to_file_mapping = {}
+        file_to_function_mapping: Dict[str, Set[int]] = defaultdict(set)
+        to_visit = deque(self.callgraph.functions.keys())
+        while to_visit:
+            function_name = to_visit.popleft()
+            if function_name in visited_functions:
+                continue
+            visited_functions.add(function_name)
+            if function_name not in function_to_file_mapping:
+                function_to_file_mapping[function_name] = self.next_function_file()
+            function_file = function_to_file_mapping[function_name]
+            file_to_function_mapping[function_file].add(function_name)
+            for target in self.callgraph.direct_branch_targets_for_function(function_name):
+                if target in self.callgraph.functions and target not in visited_functions:
+                    function_to_file_mapping[target] = function_file
+                    to_visit.appendleft(target)
+        return file_to_function_mapping
+
+    def next_function_file(self):
+        current_len = len(self.function_files)
+        new_file = f'{current_len}.c'
+        self.function_files.add(new_file)
+        return new_file
+
+    def _group_functions_with_splits(self, splits: int,
```

Comment:
Can we make this function name more clear about the fact that it's consuming the result of `_group_functions_without_splits`, and/or rename the eariler function? Right now, it reads as if the only difference is with or without splits, when actually they're doing very different things. How about: `_group_functions_by_control_flow`, and `_split_function_groups`?

Reply:
Done.